source/extensions/filters/common/rbac/engine_impl.h

```diff
@@ -13,23 +13,20 @@ namespace RBAC {
 
 class RoleBasedAccessControlEngineImpl : public RoleBasedAccessControlEngine {
 public:
-  RoleBasedAccessControlEngineImpl(const envoy::config::filter::http::rbac::v2::RBAC& config,
-                                   bool disabled);
-  RoleBasedAccessControlEngineImpl(
-      const envoy::config::filter::http::rbac::v2::RBACPerRoute& per_route_config);
+  RoleBasedAccessControlEngineImpl(const envoy::config::rbac::v2alpha::RBAC& rules);
 
   bool allowed(const Network::Connection& connection,
                const Envoy::Http::HeaderMap& headers) const override;
 
 private:
-  // Indicates that the engine will not evaluate an action and just return true for calls to
-  // allowed. This value is only set by route-local configuration.
-  const bool engine_disabled_;
-
-  // Indicates the behavior to take if a policy matches an action.
   const bool allowed_if_matched_;
 
   std::vector<PolicyMatcher> policies_;
+
+  // Indicates that the engine will not evaluate an action and just return true
+  // for calls to allowed, will set to true if empty RBAC rules is passed
+  // in constructor.
+  bool engine_disabled_;
```

Comment:
I don't think you need this variable, it is equivalent with `policies_.empty()`, no?

Reply:
no exactly the same, it's equivalent to rules not set(indicates authN not enabled), uses member variable here so don't need to check in each allow call.

Reply:
Right, I meant if you change first 3 lines of `allowed` to:

```
if (policies_.emtpy()) {
  return !allowed_if_matched_;
}
```

Then you don't need this variable.

btw, I don't think either version of the optimization really improve performance, so I'd rather remove them.

Reply:
if remove engine_disabled_, we need to put rules_ as a member variable instead, and allow will start as below, either way looks good to me([rules_](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/rbac/v2alpha/rbac.proto#L48) = action + policies)
```
envoy::config::rbac::v2alpha::RBAC empty;
if (Protobuf::util::MessageDifferencer::Equivalent(empty, rules_)) {
   return true
}
```

Reply:
`allow_if_matched_` is derived from action, so the suggestion above is correct. You don't have to use MessageDifferencer at all.

Reply:
moved checking whether rbac is enabled(by checking whether rules is set) logic from engine to filter(for routespecific as well).